### PR TITLE
CSHARP-3045: Deprecate geoHaystack and geoSearch.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**

### CODE REVIEW

File: src/MongoDB.Driver.Legacy/Builders/IndexKeysBuilder.cs [hunk 1]
```diff
@@ -66,6 +66,7 @@ public static IndexKeysBuilder GeoSpatial(string name)
         /// </summary>
         /// <param name="name">The key name.</param>
         /// <returns>The builder (so method calls can be chained).</returns>
+        [Obsolete("The geoHaystack was deprecated in server version 4.4.")]
```
Comment (rstam) [inline src/MongoDB.Driver.Legacy/Builders/IndexKeysBuilder.cs]: Replace all Obsolete messages with this one that reads better:

    [Obsolete("GeoHaystack indexes were deprecated in server version 4.4.")]

File: src/MongoDB.Driver.Core/Core/Operations/GeoSearchOperation.cs [hunk 2]
```diff
@@ -30,6 +30,7 @@ namespace MongoDB.Driver.Core.Operations
     /// Represents the geoSearch command.
     /// </summary>
     /// <typeparam name="TResult">The type of the result.</typeparam>
+    [Obsolete("The geoSearch was deprecated in server version 4.4.")]
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Operations/GeoSearchOperation.cs]: Replace this Obsolete message with this one that reads better:

    [Obsolete("The geoSearch command was deprecated in server version 4.4.")]

See next comment for a rewording of all other new Obsolete message.

File: src/MongoDB.Driver.Legacy/Builders/IndexKeysBuilder.cs (hunk 1 above)
Comment (rstam) [reply src/MongoDB.Driver.Legacy/Builders/IndexKeysBuilder.cs]: TODO: Obsolete all occurrences of `BucketSize` with the same message.

BucketSize is only used by GeHaystack indexes.

**Review by rstam — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Legacy/Builders/IndexKeysBuilder.cs (hunk 1 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Legacy/Builders/IndexKeysBuilder.cs]: done

File: src/MongoDB.Driver.Core/Core/Operations/GeoSearchOperation.cs (hunk 2 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Operations/GeoSearchOperation.cs]: done

**Comment by DmitryLukyanov** [comment]
ready for review

**Review by rstam — APPROVED** [review]
LGTM

**Review by jyemin — APPROVED** [review]
LGTM

**Pull request closed without merging**